irods/test/data_obj_test.py

```diff
@@ -1517,6 +1518,66 @@ def enter_file_into_irods( session, filename, **kw_opt ):
+        root = unique_name(my_function_name(),datetime.now(),'root')
+        home = helpers.home_collection(self.sess)
+        with self.create_resc_hierarchy(root) as (_ , Leaf):
+            with self.assertRaises(DIRECT_CHILD_ACCESS):
+                self.sess.data_objects.open('{home}/disallowed_243'.format(**locals()), 'w', **{kw.RESC_NAME_KW:Leaf})
+
+    def test_data_open_on_named_resource__243(self):
+        s = self.sess
+        data = s.data_objects
+        home = helpers.home_collection(s)
+        data_name = unique_name(my_function_name(),datetime.now(),'data')
+        resc_name = unique_name(my_function_name(),datetime.now(),'resc')
+        with self.create_simple_resc(resc_name) as resc:
+            data_path = '{home}/{data_name}'.format(**locals())
+            try:
+                with data.open(data_path,'w',**{kw.RESC_NAME_KW:resc}) as f:
+                    f.write(b'abc')
+                d = data.get(data_path)
+                self.assertEqual(len(d.replicas),1)
+                self.assertEqual(d.replicas[0].resource_name, resc)
+            finally:
+                if data.exists(data_path):
+                    data.unlink(data_path, force = True)
+
+
+    def test_access_through_resc_hierarchy__243(self):
+        s = self.sess
+        data_path = '{coll}/{data}'.format(coll = helpers.home_collection(s),
+                                           data = unique_name(my_function_name(),datetime.now()))
+        try:
+            parent = s.resources.create('parent','deferred')
+            with self.create_simple_resc('resc0_243') as r0, \
+                 self.create_simple_resc('resc1_243') as r1:
+                s.resources.add_child(parent.name,r0)
+                s.resources.add_child(parent.name,r1)
+                hiers = ['parent;{0}'.format(r) for r in (r0,r1)]
+
+                # Write two different replicas. Although the writing of the second will cause the first to become
+                # stale, each replica can be independently read by specifying the resource hierarchy.
+                for hier in hiers:
+                    opts = {kw.RESC_HIER_STR_KW: hier}
+                    with s.data_objects.open(data_path, 'a', **opts) as obj_io:
+                        obj_io.seek(0)
+                        obj_io.write(hier.encode('utf-8'))   # Write different content to each replica
+
+                # Assert that we are able to read both replicas' content faithfully using the hierarchy string.
+                for hier in hiers:
+                    opts = {kw.RESC_HIER_STR_KW: hier}
+                    with s.data_objects.open(data_path, 'r', **opts) as obj_io:
+                        self.assertEqual(obj_io.read(),hier.encode('utf-8'))   # Does unique replica have unique content?
+
+                s.data_objects.unlink(data_path, force = True)
+                s.resources.remove_child(parent.name,r1)
+                s.resources.remove_child(parent.name,r0)
```

Comment:
Could be in a `finally` block?

Reply:
Yes, or - even better (or at least, less confusingly) - use an RAII-like helper class.